test/functional/test_framework/util.py

```diff
@@ -267,6 +267,13 @@ def satoshi_round(amount: Union[int, float, str], *, rounding: str) -> Decimal:
     """Rounds a Decimal amount to the nearest satoshi using the specified rounding mode."""
     return Decimal(amount).quantize(SATOSHI_PRECISION, rounding=rounding)
 
+def ensure_for_helper_internal(*, duration, predicate, delay=0.2):
+    time_end = time.time() + duration
+    predicate_source = "''''\n" + inspect.getsource(predicate) + "'''"
+    while time.time() < time_end:
+        if not predicate():
+            raise AssertionError("Predicate {} became false within {} seconds".format(predicate_source, duration))
+        time.sleep(delay)
```

Comment:
tiny nit: if `duration < delay`, then `delay` should be set to duration?

Reply:
Hm, just doing that would have an unexpected effect, I think. With the code unchanged if the user would today pass the parameters with delay=duration, then I think the predicate would only be checked once in the beginning and not at the end because at the second conditional check the time passed should be the runtime of predicate + the sleep of duration, so the loop wouldn't run a second time. That's not what a user expects when they pass a high delay probably. So I have done the following: If delay is 0, negative or higher than duration, we set it equal to duration. But also the predicate is now check once in the beginning explicitly and in the loop it now runs after the sleep instead of before. This means that if the user sets delay to 0 explicitly (or set a value that is weird, negative or above duration) they will have the behavior or check once in the beginning and once at the end. This should be the minimum of checks, as discussed previously.